python/pyiceberg/transforms.py

```diff
@@ -215,6 +227,243 @@ def __repr__(self) -> str:
+        return "time"
+
+    def __repr__(self) -> str:
+        return "MonthTransform()"
+
+
+class DayTransform(Transform[S, int]):
+    """Transforms a datetime value into a day value.
+
+    Example:
+        >>> transform = MonthTransform()
+        >>> transform.transform(DateType())(17501)
+        17501
+    """
+
+    __root__: Literal["day"] = Field(default="day")
+    _source_type: IcebergType = PrivateAttr()
+
+    def transform(self, source: IcebergType) -> Callable[[Optional[S]], Optional[int]]:
+        source_type = type(source)
+        if source_type == DateType:
+
+            def day_func(v):
+                return v
+
+        elif source_type in {TimestampType, TimestamptzType}:
+
+            def day_func(v):
+                return datetime.micros_to_days(v)
+
+        else:
+            raise ValueError(f"Cannot apply day transform for type: {source}")
+
+        return lambda v: day_func(v) if v else None
+
+    def can_transform(self, source: IcebergType) -> bool:
+        return type(source) in {
+            DateType,
+            TimestampType,
+            TimestamptzType,
+        }
+
+    def result_type(self, source: IcebergType) -> IcebergType:
+        return DateType()
+
+    @property
+    def preserves_order(self) -> bool:
+        return True
+
+    @property
+    def time_order(self) -> int:
+        return 1
+
+    def satisfies_order_of(self, other: Transform) -> bool:
+        return self.time_order <= other.time_order if hasattr(other, "time_order") else False
+
+    def to_human_string(self, _: IcebergType, value: Optional[S]) -> str:
+        return datetime.to_human_day(value) if isinstance(value, int) else "null"
+
+    @property
+    def dedup_name(self) -> str:
+        return "time"
+
+    def __repr__(self) -> str:
+        return "DayTransform()"
+
+
+class HourTransform(Transform[S, int]):
+    """Transforms a datetime value into a hour value.
+
+    Example:
+        >>> transform = HourTransform()
+        >>> transform.transform(TimestampType())(1512151975038194)
+        420042
+    """
+
+    __root__: Literal["hour"] = Field(default="hour")
+    _source_type: IcebergType = PrivateAttr()
+
+    def transform(self, source: IcebergType) -> Callable[[Optional[S]], Optional[int]]:
+        if type(source) in {TimestampType, TimestamptzType}:
+
+            def hour_func(v):
+                return datetime.micros_to_hours(v)
+
+        else:
+            raise ValueError(f"Cannot apply hour transform for type: {source}")
+
+        return lambda v: hour_func(v) if v else None
+
+    def can_transform(self, source: IcebergType) -> bool:
+        return type(source) in {
+            TimestampType,
+            TimestamptzType,
+        }
+
+    def result_type(self, source: IcebergType) -> IcebergType:
+        return IntegerType()
+
+    @property
+    def preserves_order(self) -> bool:
+        return True
+
+    @property
+    def time_order(self) -> int:
+        return 0
```

Comment:
Instead of using these numbers, what do you think of adding an `IntEnum` to make it a bit more readable?
```
➜  python git:(fd-add-http-code-to-error-message) ✗ python3
Python 3.9.13 (main, May 24 2022, 21:13:51) 
[Clang 13.1.6 (clang-1316.0.21.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from enum import IntEnum
>>> 
>>> class TimeResolution(IntEnum):
...     YEAR = 6
...     MONTH = 5
...     WEEK = 4
...     DAY = 3
...     HOUR = 2
...     MINUTE = 1
...     SECOND = 0
... 
>>> 
>>> TimeResolution.YEAR <= TimeResolution.MONTH
False
```

Reply:
I agree. It would be much better to have an enum for this.

Reply:
👌  updated.